### PR TITLE
IGNITE-13865 Support  DateTime as a key or value in .NET and Java

### PR DESCRIPTION
**Pull request by nizhikov**
Currently, .Net type `DateTime` can't be used as a key or value in the cache in a way that Java will read values in `java.util.Date`.

This PR fixes it with the new flag `BinaryConfiguration#ForceTimestamp`.
Currently existing `BinaryReflectiveSerializer#_forceTimestamp` becomes obsolete. 

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/Marshaller.cs [hunk 1]
```diff
@@ -950,5 +950,13 @@ private static IBinaryNameMapper GetDefaultNameMapper()
         {
             return BinaryBasicNameMapper.FullNameInstance;
         }
+
+        /// <summary>
+        /// Gets force timestamp flag value.
+        /// </summary>
+        public bool ForceTimestamp()
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/Marshaller.cs:957]: Let's convert this method to a property

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryReflectiveSerializer.cs [hunk 2]
```diff
@@ -46,6 +46,7 @@ public sealed class BinaryReflectiveSerializer : IBinarySerializer
         private bool _isInUse;
 
         /** Force timestamp flag. */
+        [Obsolete("Deprecated, use BinaryConfiguration.ForceTimestamp instead.")]
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryReflectiveSerializer.cs:49]: I don't think this deprecation is necessary. This flag is still useful.

PS `[Obsolete]` should be applied to a public property, not a private field

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryReflectiveSerializer.cs (hunk 2 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Binary/BinaryReflectiveSerializer.cs]: Reverted.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/Marshaller.cs (hunk 1 above)
Comment (nizhikov) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Binary/Marshaller.cs]: Fixed.

**Comment by nizhikov** [comment]
@ptupitsyn I fixed your comments. Please, take a look one more time.

**Review by ptupitsyn — APPROVED** [review]

**Pull request merged**